entry/src/entry_batch.rs

```diff
@@ -0,0 +1,1268 @@
+use {
+    crate::entry::Entry,
+    serde::{
+        de::{self, Visitor},
+        Deserialize, Deserializer, Serialize, Serializer,
+    },
+    solana_clock::Slot,
+    solana_hash::Hash,
+    std::fmt,
+};
+
+/// A batch of entries with optional special metadata.
+///
+/// EntryBatch enforces validation rules but allows flexible combinations:
+/// - Entries can be present with or without special data
+/// - Special data can be present with or without entries
+/// - Only restriction: entries cannot be both present and special data present simultaneously
+/// - Empty entries with no special data is allowed
+///
+/// # Serialization Format
+/// The binary format is:
+/// [8 bytes: entries_len as u64 little-endian]
+/// [variable: serialized entries using bincode, concatenated]
+/// [variable: optional VersionedSpecialEntry if present]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum EntryBatch {
+    EntryVec(Vec<Entry>),
+    Special(VersionedSpecialEntry),
+}
+
+/// A versioned wrapper around SpecialEntry for backward compatibility.
+///
+/// # Serialization Format
+/// [2 bytes: version as u16 little-endian]
+/// [variable: serialized SpecialEntry]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct VersionedSpecialEntry {
+    pub version: u16,
+    pub inner: SpecialEntry,
+}
```

Comment:
Why are certain structs versioned with a version field and others are enums?

Reply:
Can we include a description at the top of the file of how we imagine this versioning will all work together?